packages/core-contracts/contracts/token/ERC721.sol

```diff
@@ -0,0 +1,232 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "../interfaces/IERC721.sol";
+import "../interfaces/IERC721Metadata.sol";
+import "../interfaces/IERC721Receiver.sol";
+import "./ERC721Storage.sol";
+import "../utils/ContractUtil.sol";
+
+contract ERC721 is IERC721, IERC721Metadata, ERC721Storage, ContractUtil {
+    error InvalidOwner(address);
+    error InvalidOperator(address);
+    error InvalidTokenId(uint256);
+    error InvalidFrom(address);
+    error InvalidTo(address);
+    error InvalidTransferRecipient(address);
+    error Unauthorized(address);
+    error AlreadyInitialized();
+
+    function _initialize(
+        string memory tokenName,
+        string memory tokenSymbol,
+        string memory initBaseTokenURI
+    ) internal virtual {
+        ERC721Store storage store = _erc721Store();
+        if (bytes(store.name).length > 0 || bytes(store.symbol).length > 0 || bytes(store.baseTokenURI).length > 0) {
+            revert AlreadyInitialized();
+        }
+
+        store.name = tokenName;
+        store.symbol = tokenSymbol;
+        store.baseTokenURI = initBaseTokenURI;
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return
+            interfaceId == this.supportsInterface.selector || // ERC165
+            interfaceId == type(IERC721).interfaceId ||
+            interfaceId == type(IERC721Metadata).interfaceId;
+    }
+
+    function name() external view override returns (string memory) {
+        return _erc721Store().name;
+    }
+
+    function symbol() external view override returns (string memory) {
+        return _erc721Store().symbol;
+    }
+
+    function tokenURI(uint256 tokenId) external view override returns (string memory) {
+        if (!_exists(tokenId)) {
+            revert InvalidTokenId(tokenId);
+        }
+
+        string memory baseURI = _erc721Store().baseTokenURI;
+        return bytes(baseURI).length > 0 ? string(abi.encodePacked(baseURI, _toString(tokenId))) : "";
+    }
+
+    function balanceOf(address holder) public view override returns (uint) {
```

Comment:
changed wording internally: owner for holder to identify the address that holds the token right now, since for our usage, the owner will always be the system (preparing the field for next PR)